File: src/restricted-extras-qscintilla/libqscistyledtext.cpp
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qscistyledtext.h>
#include "libqscistyledtext.h"
#include "libqscistyledtext.hxx"

QsciStyledText* QsciStyledText_new(libqt_string text, int style) {
    QString text_QString = QString::fromUtf8(text.data, text.len);
    return new QsciStyledText(text_QString, static_cast<int>(style));
}

QsciStyledText* QsciStyledText_new2(libqt_string text, QsciStyle* style) {
    QString text_QString = QString::fromUtf8(text.data, text.len);
    return new QsciStyledText(text_QString, *style);
}

QsciStyledText* QsciStyledText_new3(QsciStyledText* param1) {
    return new QsciStyledText(*param1);
}

void QsciStyledText_Apply(const QsciStyledText* self, QsciScintillaBase* sci) {
    self->apply(sci);
}

libqt_string QsciStyledText_Text(const QsciStyledText* self) {
    const QString _ret = self->text();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy(_str.data, _b.data(), _str.len);
    _str.data[_str.len] = '\0';
    return _str;
}

int QsciStyledText_Style(const QsciStyledText* self) {
    return self->style();
}

void QsciStyledText_Delete(QsciStyledText* self) {
    delete self;
}

File: src/restricted-extras-qscintilla/libqscistyledtext.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciStyledText.html
pub const qscistyledtext = struct {
    /// New constructs a new QsciStyledText object.
    ///
    /// ``` text: []const u8, style: i32 ```
    pub fn New(text: []const u8, style: i32) QtC.QsciStyledText {
        const text_str = qtc.struct_libqt_string{
            .len = text.len,
            .data = @constCast(text.ptr),
        };

        return qtc.QsciStyledText_new(text_str, @intCast(style));
    }

    /// New2 constructs a new QsciStyledText object.
    ///
    /// ``` text: []const u8, style: QtC.QsciStyle ```
    pub fn New2(text: []const u8, style: ?*anyopaque) QtC.QsciStyledText {
        const text_str = qtc.struct_libqt_string{
            .len = text.len,
            .data = @constCast(text.ptr),
        };

        return qtc.QsciStyledText_new2(text_str, @ptrCast(style));
    }

    /// New3 constructs a new QsciStyledText object.
    ///
    /// ``` param1: QtC.QsciStyledText ```
    pub fn New3(param1: ?*anyopaque) QtC.QsciStyledText {
        return qtc.QsciStyledText_new3(@ptrCast(param1));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciStyledText.html)
    ///
    /// ``` self: QtC.QsciStyledText, sci: QtC.QsciScintillaBase ```
    pub fn Apply(self: ?*anyopaque, sci: ?*anyopaque) void {
        qtc.QsciStyledText_Apply(@ptrCast(self), @ptrCast(sci));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciStyledText.html)
    ///
    /// ``` self: QtC.QsciStyledText, allocator: std.mem.Allocator ```
    pub fn Text(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QsciStyledText_Text(@ptrCast(self));
        defer qtc.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qscistyledtext.Text: Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciStyledText.html)
    ///
    /// ``` self: QtC.QsciStyledText ```
    pub fn Style(self: ?*anyopaque) i32 {
        return qtc.QsciStyledText_Style(@ptrCast(self));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciStyledText.html)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QsciStyledText ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QsciStyledText_Delete(@ptrCast(self));
    }
};
